smart_open/tests/test_gcs.py

```diff
@@ -0,0 +1,659 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2019 Radim Rehurek <me@radimrehurek.com>
+#
+# This code is distributed under the terms and conditions
+# from the MIT License (MIT).
+#
+import gzip
+import inspect
+import io
+import logging
+import os
+import time
+import uuid
+import unittest
+try:
+    from unittest import mock
+except ImportError:
+    import mock
+import warnings
+
+import google.cloud
+import google.api_core.exceptions
+import six
+
+import smart_open
+
+BUCKET_NAME = 'test-smartopen-{}'.format(uuid.uuid4().hex)
+BLOB_NAME = 'test-blob'
+WRITE_BLOB_NAME = 'test-write-blob'
+DISABLE_MOCKS = os.environ.get('SO_DISABLE_MOCKS') == "1"
+
+logger = logging.getLogger(__name__)
+
+
+def ignore_resource_warnings():
+    if six.PY2:
+        return
+    warnings.filterwarnings("ignore", category=ResourceWarning, message="unclosed.*<ssl.SSLSocket.*>")  # noqa
+
+
+class FakeBucket(object):
```

Comment:
It looks like you're implementing your own mocks. Is there really no library to mock GCS? (looks like [there isn't](https://github.com/googleapis/google-cloud-python/issues/5816)).

If so, then we need to write tests for these mocks as well, to ensure they're behaving as expected. The risk is that the mock implementation is buggy in a particular way that hides bugs in our GCS tests.

Reply:
I was unable to find a mocking library for GCS. I will write tests for the mocks. Do you think these should all reside in the same file or should there be something like one file for the mocks, `gcs_utils.py` and another for the mock tests `test_gcs_utils`?

Reply:
Same file is fine.